src/recovery/store/in_mem.rs

```diff
@@ -331,58 +331,201 @@ fn write_discard_drops_key() {
 
 /// A progress store with all data in-memory.
 #[derive(Debug)]
-pub(crate) struct InMemProgress(HashMap<WorkerKey, BorderEpoch>);
+pub(crate) struct InMemProgress {
+    /// Stores the current execution.
+    ///
+    /// Defaults to `-1` so the initial call to [`resume_from`] will
+    /// generate the execution `0`. There will never be any recovery
+    /// progress data that will match execution `-1`, so as soon as we
+    /// start reading any, we'll bump to the correct execution.
+    ex: i128,
+    frontiers: HashMap<WorkerIndex, WorkerFrontier>,
+}
 
 impl InMemProgress {
-    pub(crate) fn new() -> Self {
-        Self(HashMap::new())
+    /// Init to the "default" progress of a cluster of a given
+    /// size.
+    ///
+    /// You have to specify a worker count so the initial data is
+    /// well-formed.
+    pub(crate) fn new(count: WorkerCount) -> Self {
+        let ex = Into::<i128>::into(<u64 as Timestamp>::minimum()) - 1;
```

Comment:
Am I missing something, or could this just be:
```rust
let ex = -1;
```
you are requesting the minimum value of `Timestamp<T>`   which internally calls `T::min_value()` , but since `T` is hardcoded as `u64`, this should always resolve to `0`

Reply:
I was hoping that explicitly using `Timestamp::minimum` would help explain why we want the number `-1` but perhaps the type noise here makes it confusing. I don't like the idea of having hard-coded magic numbers, but I guess that's basically what's happening here no matter what.